### PR TITLE
HttpClientFactory (Review)

### PR DESCRIPTION
**Pull request by kalyanar**
@justinedelson / @davidjgonzalez 

I have created a version based on your inputs. Please review.
Please let me know the test cases I need to write.

### CODE REVIEW

**Comment by justinedelson** [comment]
I'll do more of a code review after vacation, but generally this looks good. Just a few comments:

* is Executor thread safe? I assume not, in which case a new instance should be returned from each call to getExecutor()?
* the parameter to the various methods which return a Request should be named path, as the value gets appended to the base URL to get the full URL.
* please check which of these dependencies come OOTB (with 6.0 and 6.1 both). Anything which isn't OOTB needs to be optionally imported. And we should provide a content package in a separate project for convienence.

**Comment by kalyanar** [comment]
@justinedelson I was googling and as per https://issues.apache.org/jira/browse/HTTPCLIENT-1437, the executor is thread safe. I thought that for the same domain, one executor can be shared. Each executor allows 200 concurrent requests by default. 

I shall fix other items. I shall fix the executor threadsafe issue after your further inputs.

File: bundle/src/main/java/com/adobe/acs/commons/http/impl/HttpClientFactoryImpl.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import org.apache.http.HttpHost;
+import org.apache.http.client.fluent.Executor;
+import org.apache.http.client.fluent.Request;
+import org.apache.sling.commons.osgi.PropertiesUtil;
+
+import java.util.Map;
+
+
+@Component(
+        label = "Http Components Fluent Executor Factory",
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/http/impl/HttpClientFactoryImpl.java]: label = "ACS AEM Commons - Http Components Fluent Executor Factory",

File: bundle/src/main/java/com/adobe/acs/commons/http/impl/HttpClientFactoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        description = "Http Components Fluent Executor Factory", immediate =
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/http/impl/HttpClientFactoryImpl.java]: immediate is false by default, can leave off

File: bundle/src/main/java/com/adobe/acs/commons/http/impl/HttpClientFactoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        false,
+        metatype = false, configurationFactory = true, policy =
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/http/impl/HttpClientFactoryImpl.java]: We dont want `metatype=true` ? Seems like might be handy for developers to use. Ultimately these configs should be deployed via osgiConfigs, but does it hurt to allow config via Felix?

File: bundle/pom.xml [hunk 4]
```diff
@@ -478,6 +499,11 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
```
Comment (davidjgonzalez) [inline bundle/pom.xml]: This looks like a duplicate dependency?

File: pom.xml [hunk 5]
```diff
@@ -207,6 +207,32 @@
     <dependencyManagement>
         <dependencies>
             <dependency>
+                <groupId>org.apache.httpcomponents</groupId>
```
Comment (davidjgonzalez) [inline pom.xml]: Are all these dependencies provided by core AEM bundles? Im not sure that they are...
Ex. My AEM 6.0 cannot find locate a bundle that exports `org.apache.http.client.fluent.Request` via depfinder..

The jackson stuff seems to be exported via a AEM Social UGC bundle (TBH the way soco iterates i wouldn't be surprised if that export disappeared in a few versions; almost feels like spillage.)

File: bundle/src/main/java/com/adobe/acs/commons/http/JSONResponseHandler.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
+/*
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/http/JSONResponseHandler.java]: Any chance you can reformat using the projects formatting? It looks like youre at 80 char breaks which makes it rather hard to read.

File: bundle/src/main/java/com/adobe/acs/commons/http/impl/HttpClientFactoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    private static final String USE_SSL = "use.ssl";
+
+    @Property(label = "Disable certificate check", description = "If selected" +
+            " it will disable certificate check for the SSL connection",
+            boolValue = false)
+    private static final String DISABLE_CERTIFICATE_CHECK = "disable" +
+            ".certificate";
+
+    @Property(label = "username", description = "user name", value = "admin")
+    private static final String USERNAME = "sap.username";
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/http/impl/HttpClientFactoryImpl.java]: `sap.username` to `username` ?

File: bundle/src/main/java/com/adobe/acs/commons/http/POJOResponseHandler.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import java.nio.charset.Charset;
+
+/**
+ * Converts response to a pojo.
+ *
+ * @param <T>
+ */
+public class POJOResponseHandler<T> implements ResponseHandler<T> {
+
+    public static final Charset DEFAULT_CHARSET = Charset.forName("UTF-8");
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/http/POJOResponseHandler.java]: Where is this used?

File: bundle/src/main/java/com/adobe/acs/commons/http/POJOResponseHandler.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import java.io.InputStream;
+import java.lang.reflect.Type;
+import java.nio.charset.Charset;
+
+/**
+ * Converts response to a pojo.
+ *
+ * @param <T>
+ */
+public class POJOResponseHandler<T> implements ResponseHandler<T> {
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/http/POJOResponseHandler.java]: Should this be called `JSONToPOJOResponseHandler` ? It only supports Responses that are in a JSON format, right?

File: bundle/src/main/java/com/adobe/acs/commons/http/impl/HttpClientFactoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+@Component(
+        label = "Http Components Fluent Executor Factory",
+        description = "Http Components Fluent Executor Factory", immediate =
+        false,
+        metatype = false, configurationFactory = true, policy =
+        ConfigurationPolicy.REQUIRE)
+@Service
+public class HttpClientFactoryImpl implements HttpClientFactory {
+    @Property(label = "gateway domain", description = "gateway domain", value
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/http/impl/HttpClientFactoryImpl.java]: Thought this a bit odd that its called "Domain" when the example is an "IP" .. maybe better named "host" ?

**Comment by justinedelson** [comment]
Replaced with #730

**Pull request closed without merging**